### ISSUE TITLE
trigger templater (folder template) when create tag page.

### ISSUE THREAD

**Issue by skyue** [pjeby/tag-wrangler#76]
I have set a template for default folder. But when I create tag page, the template have not been applied.

Hope to work with templater plugin.

**Comment by pjeby** [comment]
Tag Wrangler creates files using the standard Obsidian vault.create to set their initial content.  If Templater doesn't have a way to trigger application when a file is created with existing content, then no template will be applied.  You may need to manually apply the template or configure Templater differently in some fashion.  I don't use Templater myself so can't offer any more help than that.

**Comment by j-maffe** [comment]
Templater does allow to be triggered on the creation of a new file. It just need the new file to already contain the template when created.

**Comment by pjeby** [comment]
That's not going to work then.  You'll need to instead create a template for the entire tag page and have it ask you for the tag name, rather than triggering it from Tag Wrangler.

There isn't anything special about the page Tag Wrangler creates - literally *any* page with a `#tagname` alias in the front matter will work, so if you have a way of creating a templated page at all, just make one with the tag name as an alias (with `#` at the beginning) and Tag Wrangler will recognize it as the tag page for that tag.  Just make sure that you do it like this:

```md
---
Aliases: [ "#tag/name/here"]
---
```

Or this:

```md
---
Alias: "#tag/name/here"
---
```

Or this:

```md
---
Aliases:
  - "#tag/name/here"
---
```

That is, the field must be named alias or aliases, be a string or list of strings, and the tag name MUST begin with `#` and be inside quotes.  When you view the note metadata the tag should be listed as an alias, and if you use the quick switcher to open a note, then typing # and part of the tag name should show it in the list under that alias.  So as long as you set up a template that meets these criteria, you can create tag pages Tag Wrangler will recognize even if you don't use TW itself to create them.

**Comment by j-maffe** [comment]
I do have a template that I insert whenever I open a new tag page. It took me some time but I got it to work. I think it's just that some users would find it nicer if it was done automatically.

```

<%tp.frontmatter.Aliases%> `= this.file.tags[3]`

## SubTags
```dataview 
TABLE
WHERE contains(file.aliases,"<%tp.frontmatter.Aliases%>/")
FLATTEN tags
```

## Files Grouped by Folder
```dataview 
TABLE file.etags as tags
WHERE contains(file.etags,"<%tp.frontmatter.Aliases%>") AND !contains(file.etags,"<%tp.frontmatter.Aliases%>/")
FLATTEN tags
```

## Files Grouped by Tag
```dataview
TABLE WITHOUT ID map(rows, (r) => link(r.file.link, r.title)) AS Notes ,Tag 
WHERE contains(file.etags,"<%tp.frontmatter.Aliases%>") AND !contains(file.etags,"<%tp.frontmatter.Aliases%>/")
FLATTEN file.tags as Tag 
GROUP BY Tag 
```

## Tasks
```tasks
tags include <%tp.frontmatter.Aliases%>
group by path
not done
description regex does not match /^\s*$/
short mode
```
```

**Comment by MatrixFr** [comment]
how to add this own template for tag page ?

**Comment by pjeby** [comment]
It depends on which templating plugin you're using.  If you are using Obsidian's built-in template plugin, create a note in your configured templates folder, and then when you create a new note, use the "Templates: Insert Template" command (you might want to assign it a hotkey) and select the right template.

If you're using QuickAdd, you'll create a "choice" of type "Template" or "Capture" and configure it to insert what you want, where you want it.  You'll then create a command for that choice, which you can then assign a hotkey (or add a button or menu item for it if you're using Commander), and then you can use that command to create the tag page instead of using Tag Wrangler.

If you're using Templater, well, I don't really know exactly, I've never really used it.

Basically, you will need to learn to use whatever templating tool works best for you.

**Comment by mariomui** [comment]
@pjeby Basically the oped is asking for the same functionality that note refactor has in regards to templater integration.

if that's not a specific priority its fine. I also wanted to ask before contributing to the codebase as templater integration with tag wrangler is high on my own list of priorities.

Here are the proposed solutions.
* Allow the user to specify a folder for tag pages. 
  * *when a new page is created in the folder, templater can be setup to auto trigger* 
* Allow the user to specify the template WITHIN Tag Wrangler options.
* Update the Doc with stopgap templating solutions, yours and jmaffes.

I think folder specifying solution passes the test of being a feature worthy of maintenance upkeep. (it just happens to partially address the proposed trigger templater ticket.

**Comment by pjeby** [comment]
If Templater can be made to trigger a template for notes created in a folder, the simplest thing to do would be to create tag pages by making notes in that folder, then either prompt for the tag name or let the user drag the tag from the pane and drop it into the right spot.  (The next simplest thing would be to have a template that checks *every* new note and if it's a fresh/empty tag note, to automatically move/rename/expand it and otherwise leave it alone.)

There are really only two reasons Tag Wrangler includes the ability to create tag pages at all: 1) to show what a valid page looks like, and 2) so people don't *need* another plugin to do it, if they just want the page.  But if they're going to do anything fancy, it's better they use another plugin to do it, as that allows them full customization, automation, etc. -- and I don't have to duplicate any of that code.

What I *would* be willing to add to Tag Wrangler is event(s) to allow other plugins to detect (or take over) the creation of a tag page (as requested via the menu or the alt-click of a tag).  I'd also be willing to make alt-click on a tag with no tag page trigger a creation-confirmation prompt, followed by the same event sequence.  Something like:

```ts
const tp_evt: { tag: string, file?: TFile } = { tag, file: undefined };
app.workspace.trigger("tag-page:will-create", tp_evt);
if (!tp_evt.file) {
  // current code to create file
  tp_evt.file = file;
}
app.workspace.trigger("tag-page:did-create", tp_evt);
```

So a plugin could hook `will-create` and create the file if needed, or else trap did-create to modify or move the file after the fact.

**Comment by mariomui** [comment]
I considered the proposed solutions.
* 1) create tag pages by making notes in that folder, then allow secondary action from tag-wrangler.
* 2) use a template that checks every new note and if it's a fresh/empty tag note, to automatically move/rename/expand it and otherwise leave it alone.)

I've placed two criteria, a) minimum code effort(to avoid feature creep)  
and b) ease of use (1-click) 
Neither solutions proposed accomplish constraint b

## Event Driven (push-based) Solution

Your 3rd solution, i've created as follows. The difference is the allowing of a flag called this.customCreateTagPage for the create Tag Page trigger. This removes the create Page entirely and opens up the api to the user for creating a new menu item. Code work involves an if statement, and opening up your menu to an outside party. It would probably be safer to create some open apis rather than share the entire menu. Since the main goal is to create a folder , i think just sharing the create
![revealapi2](https://github.com/pjeby/tag-wrangler/assets/29618692/44226ece-e98a-4933-815f-1590448da930)
```js
export default class TagWrangler extends Plugin {
   onMenu() {
      if (this.customCreateTagPage) {
        this.app.workspace.trigger(
          'tag-wrangler:customCreateTagPage',  
          menu,
          tagName,
          {
            createTagPage: this.createTagPage,
            search,
            query,
            isHierarchy,
            tagPage,
            Keymap,
          }
        );
        console.log('hey');
      } else if (tagPage) {
        ....
      )
```

## Pull Based Solution

The user will set a property on the plugin instance (userSetTagFolderName) , if this is detected, the path updates to the folder name. As long as the file is completely empty, Templater will automatically trigger a template specific for that folder.

![empty](https://github.com/pjeby/tag-wrangler/assets/29618692/be027fe0-3c30-474e-931f-c3421b6cec0d)
```js
  async createTagPage(tagName, newLeaf) {
    const baseName = new Tag(tagName).name.split('/').join(' ');
    const folder = this.app.fileManager.getNewFileParent(
      this.app.workspace.getActiveFile()?.path || ''
    );
    const folderName = this.userSetTagFolderName ? : '/' + this.userSetTagFolderName +'/' : '';
    const path = this.app.vault.getAvailablePath(
      folder.getParentPrefix() + folderName + baseName,
      'md'
    );
    this.openTagPage(
      await this.app.vault.create(
        path,
    );
  }
  ```
  
  The Pull Based Solution uses the least amount of code. The Push based solution is probably the BEST one if you are looking to EXTEND your api.
  
  Both solutions require setting a public property (a toggler) on the instance of tag-wrangler to activate.

**Comment by pjeby** [comment]
I'm confused - you didn't comment at all on the solution I proposed...  which is the only solution I'm currently willing to implement in Tag Wrangler.  Please review my earlier comment, specifically the code, and the paragraph before and after it.

**Comment by mariomui** [comment]
That's exactly what my push code does. Event triggers are push arch. 
Also, i must have misunderstood (as I work on a mac, which would also explain my build issues.(again my PR was more to illustrate my changes rather than an actual PR)

ALT is option, or a seperate KEYMAP mod combo press. I'm not sure I agree with the usability portion of it.  
I feel like there's a middle ground by not including any text in the new page.
https://github.com/pjeby/tag-wrangler/blob/fe64b8a96f13e66dbb6b70a9c4970c7161ec0bd2/src/plugin.js#L33
```js
this.openTagPage(await this.app.vault.create(path, '');
```
Templater will only template blank notes in any solution offered anyhow.

I don't think that the opt solution (the shell solution) fulfills the constraint of usability. 
(So the final solution is to delete code, in summary)

**Comment by pjeby** [comment]
No, that is not "exactly what your push code does".  It's not even vaguely similar.  You wrote code that has one event gated by a flag that alters the context menu; mine has no flag and triggers two events when a tag page is to be created.

In case it wasn't clear, I have offered to include the code I wrote in Tag Wrangler, if you can use it.  I am not offering to do anything else until/unless I can tell you've clearly said "yes" or "no" to that offer, and if you are saying "no" I need to understand what specifically is the problem with it.

Please reply *only* with "yes, that code will work for me" or "no and here are my *specific* reasons why I can't make it work for my workflow".  Thanks.

**Comment by mariomui** [comment]
From a user perspective, no. It does not coincide with my own needs. 

(i'm not trying to be contentious. I'm also offering to contribute. Like I said, i use a ton of your stuff. I'm basically trying to get you to write a ticket for me to contribute.)
From a developer's perspective, i only include the the flag for UI/usability. The push example could have removed it. There would be an extra CREATE new page button added into your context menu (something I did not prefer)

To be honest, I don't use the tag pages enough to be part of the consuming audience. I just really like your code.
I'm just trying to get to the same page, man.

**Comment by pjeby** [comment]
Ok, I've gone ahead and implemented my proposal, since you haven't raised any user-level issues that relate to it.  (Your most recent comment only mentions issues with ways you might change *your* proposal, so I'm not sure if you've even *read* my proposal yet.)

In any case, the new feature is rolled out and available in Tag Wrangler 0.5.9, and usage instructions can be found in the new Developer Notes section of the README, if you want to use it.  (For example, if you want to write a plugin that creates tag pages using Templater, or an addition to Templater that lets you set a template for tag page creation, etc. etc.)
